pkg/agent/handler/credential.go

```diff
@@ -0,0 +1,269 @@
+package handler
+
+import (
+	"encoding/base64"
+	"fmt"
+
+	v1 "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/api/v1"
+	mv1 "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/management/v1alpha1"
+	defs "github.com/Axway/agent-sdk/pkg/apic/definitions"
+	prov "github.com/Axway/agent-sdk/pkg/apic/provisioning"
+	"github.com/Axway/agent-sdk/pkg/util"
+	"github.com/Axway/agent-sdk/pkg/util/log"
+	"github.com/Axway/agent-sdk/pkg/watchmanager/proto"
+)
+
+const (
+	xAxwayEncrypted = "x-axway-encrypted"
+	crFinalizer     = "agent.credential.provisioned"
+)
+
+type credProv interface {
+	CredentialProvision(credentialRequest prov.CredentialRequest) (status prov.RequestStatus, credentails prov.Credential)
+	CredentialDeprovision(credentialRequest prov.CredentialRequest) (status prov.RequestStatus)
+}
+
+type credentials struct {
+	prov          credProv
+	client        client
+	encryptSchema encryptSchemaFunc
+}
+
+// encryptSchemaFunc func signature for encryptSchema
+type encryptSchemaFunc func(schema, credData map[string]interface{}, key, alg, hash string) (map[string]interface{}, error)
+
+// NewCredentialHandler creates a Handler for Credentials
+func NewCredentialHandler(prov credProv, client client) Handler {
+	return &credentials{
+		prov:          prov,
+		client:        client,
+		encryptSchema: encryptSchema,
+	}
+}
+
+// Handle processes grpc events triggered for Credentials
+func (h *credentials) Handle(action proto.Event_Type, meta *proto.EventMeta, resource *v1.ResourceInstance) error {
+	if resource.Kind != mv1.CredentialGVK().Kind || h.prov == nil || isNotStatusSubResourceUpdate(action, meta) {
+		return nil
+	}
+
+	cr := &mv1.Credential{}
+	cr.FromInstance(resource)
+
+	if ok := isStatusFound(cr.Status); !ok {
+		return nil
+	}
+
+	if ok := shouldProcessPending(cr.Status.Level, cr.Metadata.State); ok {
+		log.Tracef("credential handler - processing resource in pending status")
+		cr := h.onPending(cr)
+		return h.client.CreateSubResourceScoped(cr.ResourceMeta, cr.SubResources)
+	}
+
+	if ok := shouldProcessDeleting(cr.Status.Level, cr.Metadata.State, len(cr.Finalizers)); ok {
+		log.Tracef("credential handler - processing resource in deleting state")
+		h.onDeleting(cr)
```

Comment:
I dunno, maybe i'm  not understanding these Handle funcs... I would assume h.onDeleting returns 'cr' or error, but since you're not returning it, it would return nil

Reply:
never mind... i see that onDeleting doesn't return anything

Reply:
The handlers will post results back to central, not return anything

Reply:
yup... resolving.